src/Xrefcheck/Scan.hs

```diff
@@ -170,88 +159,84 @@ readDirectoryWith mode config scanner root =
+    scanFile c = (c,) <$> scanner c
 
-    isIgnored :: FilePath -> Bool
+    isIgnored :: CanonicalPath -> Bool
     isIgnored = matchesGlobPatterns root $ ecIgnore config
 
-    -- Strip leading "." and trailing "/"
-    location :: FilePath
-    location =
-      if root `equalFilePath` "."
-        then ""
-        else dropTrailingPathSeparator root
-
 scanRepo
   :: MonadIO m
   => ScanPolicy -> Rewrite -> FormatsSupport -> ExclusionConfig -> FilePath -> m ScanResult
 scanRepo scanMode rw formatsSupport config root = do
   putTextRewrite rw "Scanning repository..."
 
-  when (not $ isDirectory root) $
+  liftIO $ whenM (not <$> doesDirectoryExist root) $
     die $ "Repository's root does not seem to be a directory: " <> root
 
+  canonicalRoot <- liftIO $ canonicalizePath root
+
   (errs, processedFiles) <-
     let mode = case scanMode of
           OnlyTracked -> RdmTracked
           IncludeUntracked -> RdmBothTrackedAndUtracked
     in liftIO
     $ (gatherScanErrs &&& gatherFileStatuses)
-    <$> readDirectoryWith mode config processFile root
+    <$> readDirectoryWith mode config processFile canonicalRoot
 
   notProcessedFiles <-  case scanMode of
     OnlyTracked -> liftIO $
-      readDirectoryWith RdmUntracked config (const $ pure NotAddedToGit) root
+      readDirectoryWith RdmUntracked config (const $ pure NotAddedToGit) canonicalRoot
     IncludeUntracked -> pure []
 
   let scannableNotProcessedFiles = filter (isJust . mscanner . fst) notProcessedFiles
 
   whenJust (nonEmpty $ map fst scannableNotProcessedFiles) $ \files -> hPutStrLn @Text stderr
     [int|A|
     Those files are not added by Git, so we're not scanning them:
-    #{interpolateBlockListF files}
+    #{interpolateBlockListF $ unCanonicalPath  <$> files}
     Please run "git add" before running xrefcheck or enable \
     --include-untracked CLI option to check these files.
     |]
 
-  let trackedDirs = foldMap (getDirs . fst) processedFiles
-      untrackedDirs = foldMap (getDirs . fst) notProcessedFiles
+  trackedDirs   <- liftIO $ foldMap (getDirs . fst) processedFiles
+  untrackedDirs <- liftIO $ foldMap (getDirs . fst) notProcessedFiles
   return . ScanResult errs $ RepoInfo
     { riFiles = M.fromList $ processedFiles <> notProcessedFiles
     , riDirectories = M.fromList
                       $ map (, TrackedDirectory) trackedDirs
                       <> map (, UntrackedDirectory) untrackedDirs
     }
-  where
-    mscanner :: FilePath -> Maybe ScanAction
-    mscanner = formatsSupport . takeExtension
 
-    isDirectory :: FilePath -> Bool
-    isDirectory = readingSystem . doesDirectoryExist
+  where
+    mscanner :: CanonicalPath -> Maybe ScanAction
+    mscanner = formatsSupport . takeExtension . unCanonicalPath
 
-    -- Get all directories from filepath.
-    getDirs :: FilePath -> [FilePath]
-    getDirs = scanl (</>) "" . splitDirectories . takeDirectory
+    getDirs :: CanonicalPath -> IO [CanonicalPath]
+    getDirs = mapM canonicalizePath
```

Comment:
Doing so can be very expensive, it will be run for every directory obtained from every file.

I guess in this case we can safely avoid performing actual canonicalization given that the incoming file is already canonicalized.

Reply:
Yes, I think that it could be done 'unsafely' as you said. By the way, I have noticed that this function is now including directories beyond the root due to the canonicalized input, so I will also fix it.